src/lxc/log.c

```diff
@@ -187,111 +186,23 @@ static int log_append_stderr(const struct lxc_log_appender *appender,
 	return 0;
 }
 
-static int lxc_unix_epoch_to_utc(char *buf, size_t bufsize, const struct timespec *time)
+static int lxc_unix_epoch_to_local(char *buf, size_t bufsize, const struct timespec *time)
 {
-	int64_t epoch_to_days, z, era, doe, yoe, year, doy, mp, day, month,
-	    d_in_s, hours, h_in_s, minutes, seconds;
-	char nanosec[INTTYPE_TO_STRLEN(int64_t)];
+#define MAX_NANO 999
 	int ret;
+    struct tm now_tm;
+    long nanosec = time->tv_nsec;
 
-	/*
-	 * See https://howardhinnant.github.io/date_algorithms.html for an
-	 * explanation of the algorithm used here.
-	 */
-
-	/* Convert Epoch in seconds to number of days. */
-	epoch_to_days = time->tv_sec / 86400;
-
-	/* Shift the Epoch from 1970-01-01 to 0000-03-01. */
-	z = epoch_to_days + 719468;
-
-	/*
-	 * Compute the era from the serial date by simply dividing by the number
-	 * of days in an era (146097).
-	 */
-	era = (z >= 0 ? z : z - 146096) / 146097;
-
-	/*
-	 * The day-of-era (doe) can then be found by subtracting the era number
-	 * times the number of days per era, from the serial date.
-	 */
-	doe = (z - era * 146097);
-
-	/*
-	 * From the day-of-era (doe), the year-of-era (yoe, range [0, 399]) can
-	 * be computed.
-	 */
-	yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
-
-	/* Given year-of-era, and era, one can now compute the year. */
-	year = yoe + era * 400;
-
-	/*
-	 * Also the day-of-year, again with the year beginning on Mar. 1, can be
-	 * computed from the day-of-era and year-of-era.
-	 */
-	doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
-
-	/* Given day-of-year, find the month number. */
-	mp = (5 * doy + 2) / 153;
-
-	/*
-	 * From day-of-year and month-of-year we can now easily compute
-	 * day-of-month.
-	 */
-	day = doy - (153 * mp + 2) / 5 + 1;
+    while (nanosec > MAX_NANO) {
+        nanosec /= 10;
+    }
 
-	/*
-	 * Transform the month number from the [0, 11] / [Mar, Feb] system to
-	 * the civil system: [1, 12] to find the correct month.
-	 */
-	month = mp + (mp < 10 ? 3 : -9);
-
-	/*
-	 * The algorithm assumes that a year begins on 1 March, so add 1 before
-	 * that.
-	 */
-	if (month < 3)
-		year++;
-
-	/* Transform days in the epoch to seconds. */
-	d_in_s = epoch_to_days * 86400;
+    localtime_r(&(time->tv_sec), &now_tm);
```

Comment:
Hey, thanks for the patch! But please read the comment above the UTC translation function. This has caused issues with Golang before and I doubt it has gotten better.

Reply:
Yeah, this definitely was a deliberate move, I also would expect issues on Android with bionic.

In the example in this PR, the output timestamp also looks wrong, either the month isn't zero padded or the fields are just plain wrong.

Reply:
thanks, comment! 
first, i fixed print format.
i will check UTC translation function

Reply:
i found issuse you opened:
- https://github.com/lxc/lxc/pull/1331
- https://github.com/lxc/lxd/issues/2189
- https://github.com/lxc/lxc/pull/1319

but no comment to show fixed this issuse, maybe we should close this pr.